### DISCUSSION TITLE
remove cgo

### DISCUSSION THREAD

**Discussion by ddkwork** [mappu/miqt#105] [Ideas]
use https://github.com/ebitengine/purego

**Comment by mappu** [comment]
See issue https://github.com/mappu/miqt/issues/7

Purego is not magic, it only allows dlsym/LoadLibrary of some C ABI library. But Qt is a C++ library, and the C++ ABI is compiler-defined, and function names are mangled. So, this requires either
- a custom .so/.dll for Miqt to interop between the C and C++ world, like [kitech/qt.go](https://github.com/kitech/qt.go) (you can use that instead already), or
- some crazy implementation of C++ structs, constructors using Go assembly, plus a custom runtime JIT for vtables.

**Comment by ddkwork** [comment]
```go
package main

import (
	"github.com/ebitengine/purego"
	"golang.org/x/sys/windows"
	"runtime"
	"syscall"
	"unsafe"
)


func main() {
	runtime.LockOSThread()
	a := NewApp(0, []string{"", ""})
	qWidget := NewQWidget()
	b := NewQPushButton("button", qWidget)
	b.OnPressed(func() {

	})
	b.Show()
	qWidget.Show()
	a.Exec()
}

var (
	QApplication_newProc            *syscall.Proc
	QWidget_newProc                 *syscall.Proc
	QWidget_showProc                *syscall.Proc
	QPushButton_newProc             *syscall.Proc
	QPushButton_showProc            *syscall.Proc
	QPushButton_connect_pressedProc *syscall.Proc
	QApplication_execProc           *syscall.Proc
	miqt_exec_callbackProc          *syscall.Proc
)

func init() {
	check(windows.SetDllDirectory("."))

	dll := syscall.MustLoadDLL("purego.dll")
	QApplication_newProc = dll.MustFindProc("QApplication_new")
	QWidget_newProc = dll.MustFindProc("QWidget_new")
	QWidget_showProc = dll.MustFindProc("QWidget_show")
	QPushButton_newProc = dll.MustFindProc("QPushButton_new")
	QPushButton_showProc = dll.MustFindProc("QPushButton_show")
	QPushButton_connect_pressedProc = dll.MustFindProc("QPushButton_connect_pressed")
	QApplication_execProc = dll.MustFindProc("QApplication_exec")
	miqt_exec_callbackProc = dll.MustFindProc("miqt_exec_callback")
}

type App struct {
	handle uintptr //todo add more fields from C++ class members
}

func NewApp(argc int, argv []string) *App {
	// 设置参数数量
	argcPtr := uintptr(unsafe.Pointer(&argc))

	// 为argv创建指针。如果没有参数，可以设置为nil
	var argvPtr uintptr
	if argv != nil && len(argv) > 0 {
		argvPtr = uintptr(unsafe.Pointer(&argv[0])) // 获取argv的指针
	} else {
		argvPtr = 0 // 或者使用nil
	}

	// 调用外部函数
	r1, _, err := QApplication_newProc.Call(argcPtr, argvPtr)
	check(err)
	return &App{
		handle: r1,
	}
}

func (a *App) Exec() {
	QApplication_execProc.Call(a.handle)
}
func check(err error) {
	if err != nil {
		//panic(err)
	}
}

type QWidget struct {
	handle uintptr //todo add more fields from C++ class members
}

func NewQWidget() *QWidget {
	r1, _, err := QWidget_newProc.Call()
	check(err)
	return &QWidget{
		handle: r1,
	}
}
func (w *QWidget) Show() {
	QWidget_showProc.Call(w.handle)
}

type QPushButton struct {
	handle uintptr //todo add more fields from C++ class members
}

func NewQPushButton(text string, parent *QWidget) *QPushButton {
	title := [256]byte{}
	copy(title[:], []byte(text))
	r1, _, err := QPushButton_newProc.Call(
		uintptr(unsafe.Pointer(&title)),
		uintptr(unsafe.Pointer(parent.handle)),
	)
	check(err)
	return &QPushButton{
		handle: r1,
	}
}

func (b *QPushButton) Show() {
	QPushButton_showProc.Call(b.handle)
}

func (b *QPushButton) OnPressed(callback func()) {
	return
	purego.NewCallback(callback) //todo
	QPushButton_connect_pressedProc.Call(b.handle)
}

```

![Screenshot_20241215_225357_com.realvnc.viewer.android.jpg](https://github.com/user-attachments/assets/d6a68b76-0009-47ec-9beb-e00fbd9a1300)

**Reply by mappu** [reply]
Nice! :star_struck: I'm glad it worked!

This is similar to the approach in kitech/qt.go: CGO is not needed, but an extra `.dll/.so` is needed to be distributed. To make MIQT available to users without CGO, a version of purego.dll must be precompiled for every OS platform.

There does also exist a way to remove the purego.dll file. The `.dll/.so` could be compiled as a `.lib/.a` archive, for static linking. Even without CGO, Go's internal linker can support this if the file is named `.syso` inside the repository. Then the purego library is not needed. However, Go's internal linker has some limitations with C++ code, so despite that the `.lib/.a` only exposes a C interface, I'm not sure if it will work completely. There is some documentation about this trick in https://go.dev/wiki/GcToolchainTricks#c-code-without--cgo- but it has some limitations.

If you want to add the `extern __declspec(dllexport)` for all MIQT functions, check it in [cmd/genbindings/emitcabi.go](https://github.com/mappu/miqt/blob/master/cmd/genbindings/emitcabi.go#L655). The header files are created by function `emitBindingHeader`. Running the `genbindings` program is slow on first use, as the full clang cache must be rebuilt. But, it's fast on second use.

By the way, I got some earlier emails from GitHub with your messages. Your english skill is excellent.

@rcalixte was doing some similar work to make C and Zig binding versions of MIQT, he may also be interested.

**Reply by ddkwork** [reply]
That's what I've been regretting, not sure if there will be side effects like virtual functions of cpp's class performing destruction memory operations and all that, as well as assuming that qt introduces some of the cpp stl stuff, as you mentioned in the readme with all those maps, hashes, containers, etc., but you said that it should work on Android, so I'm going to give it a try anyway. I'm very new to cpp templates and hopefully the templates inside qt won't have side effects on cabi:panic, memory leaks, synchronization of events within a trip, etc., but hopefully it will work as always.

Also, I'm not proficient in clang ast node traversal, so it would be great if you could easily add the dynamic library export prefixes for each platform to all gen_xxxx.h headers generated by miqt, so I can test them directly. You need to define the prefix syntax of the api export using macros according to each platform to make it compatible with all platforms.

**Comment by ddkwork** [comment]
Hi, still can't debug it after testing, also, not sure if it's possible to load so on android. do you have a static version of qt6.8 compiled by vs2022 on your side?

**Reply by mappu** [reply]
It's definitely possible to load `.so` on android - the android apk is really a zip file containing the Qt `.so` files. This is from the `mdoutliner` example:

> ![Screenshot_20241216_213453](https://github.com/user-attachments/assets/68552079-1390-4348-8cdf-de6a2bc2945c)

The MIQT loader already uses something like `dlopen/dlsym` to start up the Go program, so, I think purego will still work OK too.

---

> do you have a static version of qt6.8 compiled by vs2022 on your side?

I don't have it. The person at https://build-qt.fsu0413.me/zh-cn/ is usually quite good about making such builds, but, he has only Qt 6.7 so far. In my opinion, MSYS2 is easier than VS.

**Reply by ddkwork** [reply]
> It's definitely possible to load `.so` on android - the android apk is really a zip file containing the Qt `.so` files. This is from the `mdoutliner` example:
> 
> > ![Screenshot_20241216_213453](https://github.com/user-attachments/assets/68552079-1390-4348-8cdf-de6a2bc2945c)
> 
> The MIQT loader already uses something like `dlopen/dlsym` to start up the Go program, so, I think purego will still work OK too.
> 
> ---
> 
> > do you have a static version of qt6.8 compiled by vs2022 on your side?
> 
> I don't have it. The person at https://build-qt.fsu0413.me/zh-cn/ is usually quite good about making such builds, but, he has only Qt 6.7 so far. In my opinion, MSYS2 is easier than VS.

I have tried to compile miqt demo under msys2 and it can be compiled successfully. As you mentioned, the first compilation is very time consuming, but when I copied the qt lib downloaded from the msys2 environment to a directory on Windows 11 to test it, neither using clion+cmake to test the cpp demo nor Goland to compile the miqt demo succeeded, suggesting that all the dependencies such as zlib dependencies don't seem to be embedded in the qt lib.into the qt lib. Finally I decided to try the purego approach, which eliminates the need for a pkg config, c and c++ compilers, and compiles very quickly, but when I tested for debugging with Goland, the plugin directory for the dynamic libraries for the platform was not initialized, and adding the directory to the path of the system environment variable didn't help. In the end I think the only way to debug is to link the static version of qt. By the way, the gcc compiled qt lib works fine in msys2, but when copying the compiled miqt demo binary to run in win11, it gives me an error about the missing libc++ dynamic library, and then I found the dll and copied it to the demo directory to no avail. But the qt compiled by vs2022 works fine.

I feel that the static version of qt library+purego mode of vs2022 loads the cabi dynamic library exported by miqt, so that I can get the best development experience, and the debugging function should also work properly.

**Reply by mappu** [reply]
Yes, for the dynamic-linked MSYS2 build, many more dll files must be copied. You can use `ldd` or [depends](https://github.com/lucasg/Dependencies) or [debug/pe](https://pkg.go.dev/debug/pe#File.ImportedLibraries) to find all the required files.

It's really a lot:

![Screenshot_20241217_180258](https://github.com/user-attachments/assets/b6680a6c-470e-4ccd-8f04-d940d74ca4c8)

But, many of them are part of Windows, so copying them is not required. It's likely this is why your VS2022 build did not require any additional dll files: because the VS2022 runtime was installed at a systemwide level.

Additionally to the imports declared in the EXE file and visible to ldd, Qt also requires to load its own plugin files (e.g. `platforms/qwindows.dll`). These are loaded dynamically at runtime and are invisible to ldd. This applies to both MSYS2 Qt dynamic builds and VS2022 Qt dynamic builds.

Qt C++ has the "windeployqt" tool that can find all such dll files automatically, but I think for Miqt it must be discovered manually.

A statically linked Qt is much simpler. But because Qt is LGPL, static linking would require to make the application open-source. This can be unacceptable for some cases, so Miqt must support both ways. Although, statically linking the binding ("purego.lib / purego.a") and dynamically linking the remaining Qt libraries is acceptable.

**Comment by ddkwork** [comment]
Finally ready for debugging, need static qt library
![Screenshot_20241217_011637_com.realvnc.viewer.android.jpg](https://github.com/user-attachments/assets/7d797162-f4d7-450d-9376-2521432cce6c)

**Comment by ddkwork** [comment]
i forked a rep for make static qt 6.9 use action
https://github.com/ddkwork/qt6.9

The next step is to add static lib compilation for android, I don't know how to implement this inside action. The last step is to copy miqt's c abi bindings to generate code to implement dynamic libraries in pre-compiled c abi format for all platforms inside action.

**Comment by ddkwork** [comment]
Hello, I have a question for you.
When I try to change the code related to clang ast, I realized that the qt library on Windows doesn't have a pkg-config configuration file, it only generates a .pc file if it is compiled by the msys2 environment, or by the gcc compiler, or by the Linux or macos environment.

All the testing I've done so far has been under Windows, is there a way to use parse cmake or something else instead of pkg config to parse the compilation parameters, I'm guessing it should parse out the header file directory, compilation parameters and lib directory which are the configurations passed to clang to dump ast.

I'll try copying the .pc file of the qt lib under Linux first.

For now I need to get bindgen running before I can further change the header file export api operation.

If you have time to spare, please check every one of my commits, thanks.
https://github.com/ddkwork/miqt

**Reply by mappu** [reply]
The pkg-config `.pc` file is only used to create the needed compile flags arguments, so that Clang can resolve `#include <QWidget>` to the filepath on disk.
- It's really a simple file, you can make it by hand for Windows. There are some examples in the repo.
- Or, you can delete this part of the genbindings and replace it with your explicit arguments. In `config-libraries.go`, delete `pkgConfigCflags("Qt5Widgets")`, replace it with: `-DQT_WIDGETS_LIB -I/usr/include/x86_64-linux-gnu/qt5/QtWidgets -I/usr/include/x86_64-linux-gnu/qt5 -I/usr/include/x86_64-linux-gnu/qt5/QtCore -DQT_GUI_LIB -I/usr/include/x86_64-linux-gnu/qt5/QtGui -DQT_CORE_LIB`, well, you should use some Windows paths instead of Linux paths.

The genbindings program still looks only at Clang's AST, so you need the Clang compiler for this part.

If you have any difficulty running genbindings, a possible solution is to use the Docker container with all dependencies ready: https://github.com/mappu/miqt/blob/master/docker/genbindings.Dockerfile . This is the same container that runs in CI. It installs many extra things for all Qt 5 and Qt 6 libraries, but, the most important part is only `qt6-base-dev` and `clang`.

I run genbindings from Linux, in this way. Then afterwards I make sure that the result still works OK on Windows, using only the cross-platform parts of Qt. If you start from Windows, some `#define` may be resolved differently. It might have a small change, but it should work mostly the same.

It's a holiday season, so I can check the repository only after a few weeks. I'm really interested to see your work,

**Reply by ddkwork** [reply]
> The pkg-config file is only used to create the needed compile flags arguments, so that Clang can resolve to the filepath on disk.`.pc``#include <QWidget>`
> 
> * It's really a simple file, you can make it by hand for Windows. There are some examples in the repo.
> * Or, you can delete this part of the genbindings and replace it with your explicit arguments. In , delete , replace it with: , well, you should use some Windows paths instead of Linux paths.`config-libraries.go``pkgConfigCflags("Qt5Widgets")``-DQT_WIDGETS_LIB -I/usr/include/x86_64-linux-gnu/qt5/QtWidgets -I/usr/include/x86_64-linux-gnu/qt5 -I/usr/include/x86_64-linux-gnu/qt5/QtCore -DQT_GUI_LIB -I/usr/include/x86_64-linux-gnu/qt5/QtGui -DQT_CORE_LIB`
> 
> The genbindings program still looks only at Clang's AST, so you need the Clang compiler for this part.
> 
> If you have any difficulty running genbindings, a possible solution is to use the Docker container with all dependencies ready: https://github.com/mappu/miqt/blob/master/docker/genbindings.Dockerfile . This is the same container that runs in CI. It installs many extra things for all Qt 5 and Qt 6 libraries, but, the most important part is only and .`qt6-base-dev``clang`
> 
> I run genbindings from Linux, in this way. Then afterwards I make sure that the result still works OK on Windows, using only the cross-platform parts of Qt. If you start from Windows, some may be resolved differently. It might have a small change, but it should work mostly the same.`#define`
> 
> It's a holiday season, so I can check the repository only after a few weeks. I'm really interested to see your work,

thx, i will try.
